src/LinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -98,12 +99,80 @@ void UprobesUnwindingVisitor::visit(CallchainSamplePerfEvent* event) {
   CHECK(listener_ != nullptr);
 
   if (current_maps_ == nullptr) {
+    if (unwind_error_counter_ != nullptr) {
+      ++(*unwind_error_counter_);
+    }
     return;
   }
 
-  // TODO(b/179976268): When a sample falls on the first (push rbp) or second (mov rbp,rsp)
-  //  instruction of the current function, frame-pointer unwinding skips the caller's frame,
-  //  because rbp hasn't yet been updated to rsp. Drop the sample in this case?
+  // Callstacks with only two frames (the first is in the kernel, the second is the sampled address)
+  // are are unwinding errors.
+  // Note that this doesn't exclude samples inside the main function of any thread as the main
+  // function is never the outermost frame. For example, for the main thread the outermost function
+  // is _start, followed by __libc_start_main. For other threads, the outermost function is clone.
+  if (event->GetCallchainSize() <= 2) {
+    if (unwind_error_counter_ != nullptr) {
+      ++(*unwind_error_counter_);
+    }
+    return;
+  }
+
+  // The leaf function is not guaranteed to have the frame pointer for all our targets. Though, we
+  // assume that $rbp remains untouched by the leaf functions, such that we can rely on
+  // perf_event_open to give us "almost" correct callstacks (the caller of the leaf function will be
+  // missing). We can sanity check this assumption by checking if the callstack only contains
+  // executable code.
+  // TODO(b/187690455): As soon as we actually have frame pointers in all non-leaf functions, we can
+  //  "discard" the sample. Till that point, this check will always fail, as libc does not have
```

Comment:
"as" would actually have to be two "as"s, or the sentence is broken.
I suggest, "Till that point this check will always fail, because the caller of __libc_start_main will be invalid since libc doesn't have frame-pointers".